test/test_prototype_transforms_functional.py

```diff
@@ -128,82 +145,184 @@ def make_one_hot_label(*args, **kwargs):
+            else:
+                return str(value)
+
+        return ", ".join(
+            itertools.chain(
+                [format(arg) for arg in self.args],
+                [f"{param}={format(kwarg)}" for param, kwarg in self.kwargs.items()],
+            )
+        )
 
-class FunctionalInfo:
-    def __init__(self, name, *, sample_inputs_fn):
-        self.name = name
-        self.functional = getattr(F, name)
-        self._sample_inputs_fn = sample_inputs_fn
 
-    def sample_inputs(self):
-        yield from self._sample_inputs_fn()
+class FunctionalInfo:
+    """
+    Args:
+        functional:
+        sample_inputs_fn:
+        reference:
+        reference_inputs_fn:
+        **closeness_kwargs:
+    """
+
+    def __init__(
+        self,
+        functional,
+        *,
+        sample_inputs_fn,
+        reference=None,
+        reference_inputs_fn=None,
+        **closeness_kwargs,
+    ):
+        self.functional = functional
+        # smoke test that should hit all valid code paths
+        self.sample_inputs_fn = sample_inputs_fn
+        self.reference = reference
+        self.reference_inputs_fn = reference_inputs_fn or sample_inputs_fn
+        self.closeness_kwargs = closeness_kwargs
+
+    def __str__(self):
+        return self.functional.__name__
+
+
+def pil_reference_wrapper(pil_functional):
+    def wrapper(image_tensor, *other_args, **kwargs):
+        if image_tensor.device.type != "cpu":
+            raise pytest.UsageError("ADDME")
+        elif image_tensor.ndim > 3:
+            raise pytest.UsageError("ADDME")
+
+        try:
+            data = image_tensor.permute(1, 2, 0)
+            if data.shape[-1] == 1:
+                data.squeeze_(-1)
+            image_pil = PIL.Image.fromarray(
+                data.numpy(), mode=_COLOR_SPACE_TO_PIL_MODE.get(image_tensor.color_space, None)
+            )
+        except Exception as error:
+            raise pytest.UsageError("Converting image tensor to PIL failed with the error above.") from error
 
-    def __call__(self, *args, **kwargs):
-        if len(args) == 1 and not kwargs and isinstance(args[0], SampleInput):
-            sample_input = args[0]
-            return self.functional(*sample_input.args, **sample_input.kwargs)
+        return pil_functional(image_pil, *other_args, **kwargs)
 
-        return self.functional(*args, **kwargs)
+    return wrapper
 
 
 FUNCTIONAL_INFOS = []
 
 
-def register_kernel_info_from_sample_inputs_fn(sample_inputs_fn):
-    FUNCTIONAL_INFOS.append(FunctionalInfo(sample_inputs_fn.__name__, sample_inputs_fn=sample_inputs_fn))
-    return sample_inputs_fn
+def sample_inputs_horizontal_flip_image_tensor(device):
+    for image in make_images(sizes=[DEFAULT_IMAGE_SIZE], device=device, dtypes=[torch.float32]):
+        yield SampleInput(image)
 
 
-@register_kernel_info_from_sample_inputs_fn
-def horizontal_flip_image_tensor():
-    for image in make_images():
+def reference_inputs_horizontal_flip_image_tensor():
```

Comment:
I was expecting to see here as input args: input tensor and params (for other ops). Seems like to reference output is generated no matter the input...
I maybe missing something here.

If we need to generate sample inputs, can't we refactor this part between `def sample_inputs_horizontal_flip_image_tensor` and `reference_inputs_horizontal_flip_image_tensor` and while configuring `FunctionalInfo`, developer just specify reference function working on a single input ?

Reply:
> I was expecting to see here as input args: input tensor and params (for other ops). Seems like to reference output is generated no matter the input...
I maybe missing something here.

The `reference_inputs_fn` is not the same as the reference. The former just creates the inputs that are used for testing against the reference. It is different from the `sample_inputs_fn` in the sense that it can test a more comprehensive set of inputs with regard to their values. `sample_inputs_fn` basically only checks all valid code paths.

I'm marking this as resolved since there was quite a few refactoring steps since this comment. If this is not clear in the current state, feel free to ask again.